spyder/utils/code_analysis/client.py

```diff
@@ -218,16 +226,45 @@ def __send_request(self, request):
         self.request_seq += 1
 
 
+class TerminateSignal(Exception):
+    """Terminal exception descriptor."""
+    pass
+
+
+class SignalManager:
+    """Manage and intercept SIGTERM and SIGKILL signals."""
+
+    def __init__(self):
+        self.original_sigint = signal.getsignal(signal.SIGINT)
+        self.original_sigterm = signal.getsignal(signal.SIGTERM)
+        signal.signal(signal.SIGINT, self.exit_gracefully)
+        signal.signal(signal.SIGTERM, self.exit_gracefully)
+
+    def exit_gracefully(self, signum, frame):
+        LOGGER.info('Termination signal ({}) captured, '
+                    'initiating exit sequence'.format(signum))
+        raise TerminateSignal("Exit process!")
+
+    def restore(self):
+        signal.signal(signal.SIGINT, self.original_sigint)
+        signal.signal(signal.SIGTERM, self.original_sigterm)
+
+
 if __name__ == '__main__':
+    sig_manager = SignalManager()
     client = LanguageServerClient(host=args.server_host,
                                   port=args.server_port,
                                   workspace=args.folder,
                                   zmq_port=args.zmq_port,
                                   use_external_server=args.external_server,
                                   server=args.server,
                                   server_args=unknownargs)
+    client.start()
     try:
-        client.start()
-    except KeyboardInterrupt:
-        client.stop()
-        sys.exit(0)
+        while True:
+            client.listen()
+    except TerminateSignal:
+        pass
+    client.stop()
+    sig_manager.restore()
+    os.kill(os.getpid(), signal.SIGTERM)
```

Comment:
This doesn't work on Windows, i.e. processes can't be killed this way there. If you want a cross-platform way of doing this, please take a look at `psutil.kill`.